omc3/harpy/clean.py

```diff
@@ -173,67 +174,55 @@ def _resync_bpms(harpy_input, bpm_data, model):
+    Requiring K singular values from MxN matrix results in matrices sized: ((M,K) x diag(K) x (K,N))
 
     Args:
-        bpm_data: matrix to be decomposed
-        num_singular_values: input options object that contains
+        matrix: matrix to be decomposed
+        num_singular_values: Required number of singular values for reconstruction
         dominance_limit: limit on SVD dominance
+        num_iter: maximal number of iteration to remove elements and renormalise matrices
 
     Returns:
-        An indexed DataFrame of U matrix, product of S and V^T martices,
-            mean of original matrix and U matrix mask for cleaned elements
+        An indexed DataFrame of U matrix (M,K),
+        product of S and V^T martices (diag(K).x(K,N)),
+        and U matrix mask for cleaned elements
     """
-    bpm_data_mean = bpm_data.to_numpy().mean()
-    u_mat, svt_mat, u_mat_mask = _get_decomposition(bpm_data - bpm_data_mean, num_singular_values,
-                                                    dominance_limit=dominance_limit)
-    return pd.DataFrame(index=bpm_data.index, data=u_mat), svt_mat, bpm_data_mean, u_mat_mask
-
-
-def _get_decomposition(matrix, num, dominance_limit=None):
-    """
-    Removes noise floor
-    Requiring K singular values from MxN matrix results in matrices sized: ((MxK) x diag(K) x (K,N))
-
-    Returns:
-        U (MxK),  SVt (diag(K).(K,N)), U matrix mask for cleaned elements (same dimensions as U)
-    """
-    u_mat, s_mat, vt_mat = np.linalg.svd(matrix / np.sqrt(matrix.shape[1]), full_matrices=False)
-
-    u_mat, s_mat, u_mat_mask = _remove_dominant_elements(u_mat, s_mat, dominance_limit)
+    u_mat, s_mat, vt_mat = np.linalg.svd(matrix, full_matrices=False)
+    u_mat, s_mat, u_mat_mask = _remove_dominant_elements(u_mat, s_mat, dominance_limit, num_iter=num_iter)
 
     available = np.sum(s_mat > 0.)
-    if num > available:
+    if num_singular_values > available:
         LOGGER.warning(f"Requested more singular values than available(={available})")
-    keep = min(num, available)
-    indices = np.argsort(s_mat)[::-1][:keep]
+    keep = min(num_singular_values, available)
     LOGGER.debug(f"Number of singular values to keep: {keep}")
-    return (u_mat[:, indices],
-            np.dot(np.sqrt(matrix.shape[1]) * np.diag(s_mat[indices]), vt_mat[indices, :]),
-            u_mat_mask[:, indices])
+
+    indices = np.argsort(s_mat)[::-1][:keep]
+    return (pd.DataFrame(index=matrix.index, data=u_mat[:, indices]),
+            np.dot(np.diag(s_mat[indices]), vt_mat[indices, :]),
+            u_mat_mask[:, :int(np.max(indices))+1])
 
 
-def _remove_dominant_elements(u_mat, s_mat, dominance_limit):
+def _remove_dominant_elements(u_mat, s_mat, dominance_limit, num_iter=1):
     u_mat_mask = np.ones(u_mat.shape, dtype=bool)
     if dominance_limit is None:
         return u_mat, s_mat, u_mat_mask
     abs_dominance_limit = np.abs(dominance_limit)
```

Comment:
since signed_limit is not used anymore, this line would be redundant if options is checked in entrypoint